criu/files-reg.c

```diff
@@ -1602,6 +1608,125 @@ static int get_build_id(const int fd, const struct stat *fd_status,
 	return -1;
 }
 
+static inline void calculate_checksum_iterator_init(int *iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_EVERY)
+		*iter = 0;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST)
+		*iter = 0;
+	else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD)
+		*iter = 0;
+}
+
+static inline void calculate_checksum_iterator_next(int *iter, int *low_bound,
+							int *up_bound, size_t file_size,
+							unsigned char **file_header,
+							const int fd)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_EVERY) {
+		*iter += 1;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST) {
+		*iter += 1;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD) {
+		*iter += opts.file_validation_chksm_parameter;
+
+		/*
+		 * If the new iterator position is outside the mapped region (Bytes in the range 
+		 * low_bound and up_bound) of the file, the current region needs to be unmapped
+		 * and the next portion (At most the next 10MB) of the file needs to be mapped.
+		 */
+		if (*iter >= *up_bound && *up_bound != file_size) {
+			munmap(*file_header, *up_bound - *low_bound);
+			*iter -= *up_bound;
+			*low_bound = *up_bound;
+			*up_bound = min_t(size_t, file_size, *up_bound + CHKSM_CHUNK_SIZE);
+			*file_header = (unsigned char *) mmap(0, *up_bound - *low_bound,
```

Comment:
need to check errors of mmap

Reply:
and I would say that if you call mmap here, you need to create an initial mapping in calculate_checksum_iterator_init and to unmap in calculate_checksum_iterator_stop_condition. WDYT?

Reply:
I moved this entire block of code into calculate_checksum().